services/userlog/pkg/service/filter.go

```diff
@@ -0,0 +1,102 @@
+package service
+
+import (
+	"context"
+	"errors"
+	user "github.com/cs3org/go-cs3apis/cs3/identity/user/v1beta1"
+	"github.com/cs3org/reva/v2/pkg/events"
+	"github.com/owncloud/ocis/v2/ocis-pkg/log"
+	"github.com/owncloud/ocis/v2/ocis-pkg/middleware"
+	settingssvc "github.com/owncloud/ocis/v2/protogen/gen/ocis/services/settings/v0"
+	"github.com/owncloud/ocis/v2/services/settings/pkg/store/defaults"
+	micrometadata "go-micro.dev/v4/metadata"
+)
+
+type UserlogFilter struct {
+	log         log.Logger
+	valueClient settingssvc.ValueService
+}
+
+func NewUserlogFilter(l log.Logger, vc settingssvc.ValueService) *UserlogFilter {
+	return &UserlogFilter{log: l, valueClient: vc}
+}
+
+// Filter removes users who should not receive an in-app notifications for the event
+func (ulf UserlogFilter) Filter(ctx context.Context, event events.Event, executant *user.UserId, users []string) []string {
+	filteredUsers := ulf.filterExecutant(users, executant)
+	return ulf.filterUsersBySettings(ctx, filteredUsers, event)
+}
+
+// filterExecutant removes the executant
+func (ulf UserlogFilter) filterExecutant(users []string, executant *user.UserId) []string {
+	var filteredUsers []string
+	for _, u := range users {
+		if u != executant.GetOpaqueId() {
+			filteredUsers = append(filteredUsers, u)
+		}
+	}
+	return filteredUsers
+}
+
+// filterUsersBySettings removes users who have disabled in-app notifications for the event
+func (ulf UserlogFilter) filterUsersBySettings(ctx context.Context, users []string, event events.Event) []string {
+	var filteredUsers []string
+	var settingsId string
+	// map type to settings key
+	switch event.Event.(type) {
+	case events.ShareCreated:
+		settingsId = defaults.SettingUUIDProfileEventShareCreated
+	case events.ShareRemoved:
+		settingsId = defaults.SettingUUIDProfileEventShareRemoved
+	case events.ShareExpired:
+		settingsId = defaults.SettingUUIDProfileEventShareExpired
+	case events.SpaceShared:
+		settingsId = defaults.SettingUUIDProfileEventSpaceShared
+	case events.SpaceUnshared:
+		settingsId = defaults.SettingUUIDProfileEventSpaceUnshared
+	case events.SpaceMembershipExpired:
+		settingsId = defaults.SettingUUIDProfileEventSpaceMembershipExpired
+	case events.SpaceDisabled:
+		settingsId = defaults.SettingUUIDProfileEventSpaceDisabled
+	case events.SpaceDeleted:
+		settingsId = defaults.SettingUUIDProfileEventSpaceDeleted
+	default:
+		// event that cannot be disabled
+		return users
+	}
+
+	for _, u := range users {
+		enabled, err := getSetting(ctx, ulf.valueClient, u, settingsId)
+		if err != nil {
+			ulf.log.Error().Err(err).Msg("cannot get user event setting")
```

Comment:
Please add the user ID and setting ID log.Error().Err(err).Str("userid", userid).Msg...